cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityManagementService.java

```diff
@@ -0,0 +1,136 @@
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * Service for managing capabilities. Responsible for periodically reading capability configurations and applying them.
+ */
+public class CapabilityManagementService extends AbstractScheduledService {
+
+  private CConfiguration cConf;
+  private Map<CapabilityActionType, CapabilityActionExecutor> capabilityExecutorMap;
+  private final long scheduleInterval;
+  private CapabilityManager capabilityManager;
+  private static final Gson GSON = new Gson();
+  private static final Logger LOG = LoggerFactory.getLogger(CapabilityManagementService.class);
+
+  @Inject
+  public CapabilityManagementService(CConfiguration cConf, CapabilityManager capabilityManager,
+                                     Map<CapabilityActionType, CapabilityActionExecutor> capabilityExecutorMap) {
+    this.cConf = cConf;
+    this.capabilityExecutorMap = capabilityExecutorMap;
+    this.capabilityManager = capabilityManager;
+    scheduleInterval = cConf.getLong(Constants.Capability.CAPABILITY_SCAN_INTERVAL, 1L);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    //Subsequent executions will not run concurrently (scheduler guarantee),
+    // so need not worry about scan already in progress.
+    LOG.debug("Scanning capability config directory");
+    try {
+      scanConfigDirectory();
+    } catch (Exception exception) {
+      //If exception is thrown, no more executions will happen.
+      LOG.error("Error while scanning capability config directory ", exception);
+    }
+  }
+
+  private void scanConfigDirectory() throws FileNotFoundException {
+    File srcDir = new File(cConf.get(Constants.Capability.CAPABILITY_CONFIG_EXTERNAL_DIR));
+    File destDir = new File(cConf.get(Constants.Capability.CAPABILITY_CONFIG_DIR));
+    Set<File> toDelete = diffAndCopy(srcDir, destDir);
+    //TODO should each config run in its own thread ?
+    for (File configFile : DirUtils.listFiles(destDir)) {
+      CapabilityConfig capabilityConfig = parseConfig(configFile);
+      if (toDelete.contains(configFile)) {
+        handleDelete(capabilityConfig, configFile);
+        continue;
+      }
+      capabilityExecutorMap.get(capabilityConfig.getType()).execute(capabilityConfig);
+    }
+  }
+
+  private void handleDelete(CapabilityConfig capabilityConfig, File configFile) {
+    capabilityExecutorMap.get(CapabilityActionType.DELETE).execute(capabilityConfig);
+    String capability = capabilityConfig.getCapability();
+    if (capabilityManager.isCapabilityPresent(capability)) {
+      return;
+    }
+    //capability deleted. Remove the file
+    LOG.debug("Cleaning up the config file {} after removing capability {}", configFile, capability);
+    configFile.delete();
+  }
+
+  private Set<File> diffAndCopy(File srcDir, File destDir) {
+    List<File> srcFiles = DirUtils.listFiles(srcDir);
+    //copy the new files and replace
+    srcFiles.stream().
```

Comment:
Use a simple for-loop is better in this case. Wrapping all IOException as RuntimeException is bad.